python/feathub/common/configuration.py

```diff
@@ -0,0 +1,194 @@
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from enum import Enum
+from typing import TypeVar, Generic, Optional, Dict, Type, Any, List
+
+import pandas as pd
+
+from feathub.common.exceptions import FeathubConfigurationException
+
+T = TypeVar("T")
+
+
+# TODO: Generate documentation of all the defined configuration automatically.
+class FeathubConfiguration:
+    """
+    FeathubConfiguration holds all the configurations.
+    """
+
+    def __init__(self) -> None:
+        self.config: Dict[str, Any] = {}
+
+    @staticmethod
+    def from_dict(config_dict: Dict) -> "FeathubConfiguration":
+        config = FeathubConfiguration()
+        json_normalized = pd.json_normalize(config_dict).to_dict(orient="records")
+        if len(json_normalized) >= 1:
+            [config.config] = json_normalized
+        else:
+            config.config = {}
+        return config
+
+    def get(self, config_option: "ConfigOption[T]") -> T:
+        return config_option.get_value(self.config)
+
+    def __repr__(self) -> str:
+        return self.config.__repr__()
+
+    def __str__(self) -> str:
+        return self.config.__str__()
+
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, self.__class__) and self.config.__eq__(other.config)
+
+
+class ConfigOption(Generic[T]):
+    """
+    ConfigOption describes a configuration parameter.
+
+    It encapsulates the configuration key, description of the configuration, data type
+    of the value and an optional default value for the configuration parameter.
+    """
+
+    def __init__(
+        self,
+        *,
+        key: str,
+        description: str,
+        value_type: Type[T],
+        default_value: Optional[T],
+    ):
+        """
+        :param key: The key of the configuration.
+        :param description: The description of the configuration.
+        :param value_type: The python type of the value of the configuration.
+        :param default_value: The default value of the configuration.
+        """
+        self.key = key
+        self.default_value = default_value
+        self.description = description
+        self.value_type = value_type
+
+    def get_value(self, config: Dict) -> T:
+        value = config.get(self.key, self.default_value)
+        if not isinstance(value, self.value_type):
+            raise FeathubConfigurationException(
+                f"Configuration type error: {self.key} expects type {self.value_type} "
+                f"but got {type(value)}."
+            )
+        return value
+
+
+ENUM_T = TypeVar("ENUM_T", bound=Enum)
+
+
+class EnumConfigOption(ConfigOption[ENUM_T]):
```

Comment:
It seems better to remove this class and instead support validator as part of config definition.

Same for EnumListConfigOption.